linkerd/trace-context/src/propagation.rs

```diff
@@ -42,16 +47,111 @@ impl TraceContext {
+    // Attempt to parse as w3c first since it's the newest interface in
+    // distributed tracing ecosystem
+    unpack_w3c_trace_context(request)
+        .or_else(|| unpack_grpc_trace_context(request))
+        .or_else(|| unpack_http_trace_context(request))
 }
 
 // Generates a new span id, writes it to the request in the appropriate
 // propagation format and returns the generated span id.
 pub fn increment_span_id<B>(request: &mut http::Request<B>, context: &TraceContext) -> Id {
     match context.propagation {
-        Propagation::Grpc => increment_grpc_span_id(request, context),
-        Propagation::Http => increment_http_span_id(request),
+        Propagation::B3Grpc => increment_grpc_span_id(request, context),
+        Propagation::B3Http => increment_http_span_id(request),
+        Propagation::W3CHttp => {
+            increment_w3c_http_span_id(request, context.trace_id.clone(), context.flags.clone())
+        }
+    }
+}
+
+fn unpack_w3c_trace_context<B>(request: &http::Request<B>) -> Option<TraceContext> {
+    get_header_str(request, W3C_HTTP_TRACEPARENT).and_then(parse_w3c_context)
+}
+
+fn parse_w3c_context(header_value: &str) -> Option<TraceContext> {
+    let rest = match header_value.split_once(W3C_HEADER_VALUE_SEPARATOR) {
+        Some((version, rest)) => {
+            if version != W3C_HEADER_VERSION {
+                warn!(
+                    "Tracecontext header {} contains invalid version: {}",
+                    header_value, version
+                );
+                return None;
+            }
+            rest
+        }
+        None => {
+            warn!("Tracecontext header {} is invalid", header_value);
+            return None;
+        }
+    };
+
+    let (trace_id, rest) = if let Some((id, rest)) = parse_w3c_header_value(rest, 16) {
+        (id, rest)
+    } else {
+        warn!("Tracecontext header {} contains invalid id", header_value);
+        return None;
+    };
+
+    let (parent_id, rest) = if let Some((id, rest)) = parse_w3c_header_value(rest, 8) {
+        (id, rest)
+    } else {
+        warn!("Tracecontext header {} contains invalid id", header_value);
+        return None;
+    };
+
+    let flags = match hex::decode(rest) {
+        // If valid hex, take final bit and AND with 1. W3C only uses one bit
+        // for flags in version 00, and the bit is used to control sampling
+        Ok(decoded) => Flags(decoded[0] & 0x1),
+        // If invalid hex, do not sample trace
+        Err(e) => {
+            warn!("Failed to decode flags for header {}: {}", header_value, e);
+            Flags(0)
+        }
+    };
+
+    Some(TraceContext {
+        propagation: Propagation::W3CHttp,
+        trace_id,
+        parent_id,
+        flags,
+    })
+}
+
+fn parse_w3c_header_value(next_header: &str, pad_to: usize) -> Option<(Id, &str)> {
+    next_header
+        .split_once(W3C_HEADER_VALUE_SEPARATOR)
+        .filter(|(id, _rest)| !id.chars().all(|c| c == '0'))
```

Comment:
I added a comment there. Good catch. The spec says that if an ID is all 0s then it's invalid and the whole trace should be invalidated. It would be possible if somebody sets up a span manually, for example.